src/utilities/bmi/mass_balance.cpp

```diff
@@ -0,0 +1,178 @@
+    initialize(model, properties);
+}
+
+NgenMassBalance::NgenMassBalance() : check(false) {}
+
+NgenMassBalance::~NgenMassBalance() = default;
+
+auto NgenMassBalance::run(const ModelPtr& model, const Context& ctx) const -> expected<void, ProtocolError> {
+    bool check_step = false;
+    //if frequency was set to -1, only check at the end
+    if( frequency == -1 ){
+        if(ctx.current_time_step == ctx.total_steps){
+            check_step = true;
+        }
+    }
+    else{
+        check_step = (ctx.current_time_step % frequency) == 0;
+    }
+    if( model == nullptr ) {
+        return make_unexpected<ProtocolError>( ProtocolError(
+            Error::UNITIALIZED_MODEL,
+            "Cannot run mass balance protocol with null model."
+            )
+        );
+    }
+    if(model && is_supported && check && check_step) {
+        double mass_in, mass_out, mass_stored, mass_leaked, mass_balance;
+        model->GetValue(INPUT_MASS_NAME, &mass_in);
+        model->GetValue(OUTPUT_MASS_NAME, &mass_out);
+        model->GetValue(STORED_MASS_NAME, &mass_stored);
+        model->GetValue(LEAKED_MASS_NAME, &mass_leaked);
+        // TODO consider unit conversion if/when it becomes necessary
+        mass_balance = mass_in - mass_out - mass_stored - mass_leaked;
+        // std::stringstream ss;
+        // ss << "Mass balance for " << model->GetComponentName() <<"\n\t" <<
+        //         INPUT_MASS_NAME << "(" << mass_in <<  ") - " <<
+        //         OUTPUT_MASS_NAME << " (" << mass_out <<  ") - " << 
+        //         STORED_MASS_NAME << " (" << mass_stored << ") = " << 
+        //         mass_balance << "\n\t" << "tolerance: " << tolerance;
+        // std::cout << ss.str() << std::endl;
+        if ( std::abs(mass_balance) > tolerance ) {
+            std::stringstream ss;
+            ss << "mass_balance: "
+               << "at timestep " << std::to_string(ctx.current_time_step)
+               << " ("+ctx.timestamp+")"
+               << " at feature id " << ctx.id <<std::endl
+               << "\tMass balance check failed for " << model->GetComponentName() << "\n\t" <<
+                INPUT_MASS_NAME << "(" << mass_in <<  ") - " <<
+                OUTPUT_MASS_NAME << " (" << mass_out <<  ") - " << 
+                STORED_MASS_NAME << " (" << mass_stored <<  ") - " << 
+                LEAKED_MASS_NAME << " (" << mass_leaked << ") = " << 
+                mass_balance << "\n\t" << "tolerance: " << tolerance << std::endl;
+            return make_unexpected<ProtocolError>( ProtocolError(
+                is_fatal ? Error::PROTOCOL_ERROR : Error::PROTOCOL_WARNING,
+                ss.str()
+                )
+            );
+        }
+
+    }
+    return {};
+}
+
+auto NgenMassBalance::check_support(const ModelPtr& model) -> expected<void, ProtocolError> {
+    if (model->is_model_initialized()) {
+        double mass_var;
+        std::vector<std::string> units;
```

Comment:
Not required, but I think you could drop the collection and do the unit check below as we iterate. Alternatively, what about just switching to a `std::array` instead?

Reply:
This could simplify to two strings, `prev` and `curr` and a comparison for each iteration that the current unit is the same as the previous, and error when not.  I think that would even allow something like checking (via associative properties) if units are convertible when they don't match.  That was one case I had in mind when structuring this way, and trying to leave things in a state which allows further and/or different validation in the future.  I'm not sure its super impactful in this case, and may actually be ultimately more efficient to hold all 4 strings and iterate them as opposed to overwriting `prev`, `current` each iteration.